File: bench/cpp/ops/BenchSIFT.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchSIFTConfig.hpp"

#include <cvcuda/OpSIFT.hpp>

#include <nvbench/nvbench.cuh>

#include <string>

template<typename T>
inline void sift(nvbench::state &state, nvbench::type_list<T>)
try
{
    int3                        shape        = benchutils::GetShape<3, int3>(state.get_string("shape"));
    const benchutils::InputKind inputKind    = benchutils::GetInputKind(state.get_string("inputKind"));
    const std::string           layout       = state.get_string("layout");
    auto                        capacity     = benchutils::GetIntParam<int>(state, "maxCapacity");
    auto                        numOctLayers = benchutils::GetIntParam<int>(state, "numOctaveLayers");
    auto                        contThr      = static_cast<float>(state.get_float64("contrastThreshold"));
    auto                        edgeThr      = static_cast<float>(state.get_float64("edgeThreshold"));
    auto                        initSigma    = static_cast<float>(state.get_float64("initSigma"));

    NVCVSIFTFlagType flags;

    int3 maxShape;

    if (layout != "NHWC" && layout != "NCHW")
    {
        throw std::invalid_argument("Invalid layout = " + layout);
    }

    if (state.get_string("expandInput") == "Y")
    {
        flags    = NVCV_SIFT_USE_EXPANDED_INPUT;
        maxShape = int3{shape.z * 2, shape.y * 2, shape.x};
    }
    else if (state.get_string("expandInput") == "N")
    {
        flags    = NVCV_SIFT_USE_ORIGINAL_INPUT;
        maxShape = int3{shape.z, shape.y, shape.x};
    }
    else
    {
        throw std::invalid_argument("Invalid expandInput = " + state.get_string("expandInput"));
    }

    // Each pyramid has shape approximately (3 + L) * N * (2 HW size) * F32
    std::size_t pyrSize = (numOctLayers + 3) * shape.x * (maxShape.x * maxShape.y * 2) * sizeof(float);

    // R/W bandwidth rationale:
    // 1 read of input (U8) to build (F32) pyramids, 1 read of Gauss and 1 read of DoG pyramids
    // 1 write of Gauss and 1 write of DoG pyramids, 1 write of 4 output data
    state.add_global_memory_reads(shape.x * shape.y * shape.z * sizeof(T) + 2 * pyrSize);
    state.add_global_memory_writes(2 * pyrSize + shape.x * sizeof(int)
                                   + shape.x * capacity * (sizeof(float4) + sizeof(float3) + 128 * sizeof(T)));

    cvcuda::SIFT op(maxShape, numOctLayers);

    // clang-format off

    if (inputKind == benchutils::InputKind::Tensor)
    {
        nvcv::Tensor src = layout == "NCHW" ? nvcv::Tensor({{shape.x, 1, shape.y, shape.z}, "NCHW"}, nvcv::TYPE_U8)
                                            : nvcv::Tensor({{shape.x, shape.y, shape.z, 1}, "NHWC"}, nvcv::TYPE_U8);
        nvcv::Tensor dstC({{shape.x, capacity}, "NM"}, nvcv::TYPE_4F32);
        nvcv::Tensor dstM({{shape.x, capacity}, "NM"}, nvcv::TYPE_3F32);
        nvcv::Tensor dstD({{shape.x, capacity, 128}, "NMD"}, nvcv::TYPE_U8);
        nvcv::Tensor dstN({{shape.x}, "N"}, nvcv::TYPE_S32);

        benchutils::FillTensor<T>(src, benchutils::LcgValues<T>());

        benchutils::warmup_and_exec(state, BENCH_SIFT_WARMUP_ITERATIONS,
            [&op, &src, &dstC, &dstM, &dstD, &dstN, &numOctLayers, &contThr, &edgeThr, &initSigma, &flags](cudaStream_t s) {
                op(s, src, dstC, dstM, dstD, dstN, numOctLayers, contThr, edgeThr, initSigma, flags);
            });
    }
    else // zero and positive var shape means use ImageBatchVarShape
    {
        throw std::invalid_argument("ImageBatchVarShape not implemented for this operator");
    }
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(sift, NVBENCH_TYPE_AXES(BENCH_SIFT_TYPES))
BENCH_SIFT_AXES;

File: bench/python/ops/bench_sift.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CV-CUDA SIFT operator benchmark - Python equivalent of BenchSIFT.cpp"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_input_kind,
    parse_shape,
    create_tensor,
    create_stream_cache,
    run_benchmark,
)


def sift(state):
    """SIFT operator benchmark matching C++ BenchSIFT.cpp"""

    shape = parse_shape(state.get_string("shape"))
    input_kind = get_input_kind(state.get_string("inputKind"))
    layout = state.get_string("layout")
    expand_input = state.get_string("expandInput") == "Y"
    capacity = state.get_int64("maxCapacity")
    num_octave_layers = state.get_int64("numOctaveLayers")
    contrast_threshold = state.get_float64("contrastThreshold")
    edge_threshold = state.get_float64("edgeThreshold")
    init_sigma = state.get_float64("initSigma")
    device_id = state.get_device()

    N, H, W = shape
    if layout not in ("NHWC", "NCHW"):
        raise ValueError(f"Invalid layout = {layout}")

    input_shape = (N, 1, H, W) if layout == "NCHW" else (N, H, W, 1)

    flags = (
        cvcuda.SIFT.USE_EXPANDED_INPUT
        if expand_input
        else cvcuda.SIFT.USE_ORIGINAL_INPUT
    )

    if expand_input:
        max_w, max_h = W * 2, H * 2
    else:
        max_w, max_h = W, H

    pyr_size = (num_octave_layers + 3) * N * (max_w * max_h * 2) * 4

    dtype_size = 1
    state.add_global_memory_reads(N * H * W * dtype_size + 2 * pyr_size)
    state.add_global_memory_writes(
        2 * pyr_size + N * 4 + N * capacity * (16 + 12 + 128 * dtype_size)
    )

    if input_kind == "VarShape":
        state.skip("ImageBatchVarShape not implemented for this operator")
        return None

    src = create_tensor(input_shape, "uint8", device_id, layout=layout, fill_mode="lcg")

    feat_coords = cvcuda.Tensor((N, capacity, 4), cvcuda.Type.F32, rowalign=1)
    feat_metadata = cvcuda.Tensor((N, capacity, 3), cvcuda.Type.F32, rowalign=1)
    feat_descriptors = cvcuda.Tensor((N, capacity, 128), cvcuda.Type.U8, rowalign=1)
    num_features = cvcuda.Tensor((N, 1), cvcuda.Type.S32, rowalign=1)

    get_stream = create_stream_cache()

    def run(launch):
        stream = get_stream(launch)
        cvcuda.sift_into(
            feat_coords,
            feat_metadata,
            feat_descriptors,
            num_features,
            src,
            num_octave_layers=num_octave_layers,
            contrast_threshold=contrast_threshold,
            edge_threshold=edge_threshold,
            init_sigma=init_sigma,
            flags=flags,
            stream=stream,
        )

    return run


if __name__ == "__main__":
    run_benchmark("sift", sift)
